erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py

```diff
@@ -4,9 +4,127 @@
 import frappe
+from frappe import _
+from frappe.utils import flt, formatdate, getdate
 from frappe.model.document import Document
+from erpnext.hr.doctype.leave_allocation.leave_allocation import get_carry_forwarded_leaves
 
 class CompensatoryLeaveRequest(Document):
 	def validate_present(self):
 		pass
 
+	def validate(self):
+		self.validate_dates()
+		self.validate_overlap()
+
+	def on_submit(self):
+		if not self.leave_type:
+			frappe.throw(_("Please select a leave type to submit the request"))
+		else:
+			company = frappe.db.get_value("Employee", self.employee, "company")
+			leave_period = self.get_leave_period(company)
+			if leave_period:
+				max_leaves_allowed = frappe.db.get_value("Leave Type", self.leave_type, "max_leaves_allowed")
+				self.validate_leave_allocation_days(leave_period, max_leaves_allowed)
+				self.create_leave_allocation(leave_period, max_leaves_allowed)
+
+	def create_leave_allocation(self, leave_period, max_leaves_allowed):
+		leave_allocated = get_leave_allocation_for_period(self.employee, self.leave_type, leave_period[0].from_date, leave_period[0].to_date)
+		leave_alloc_balance = max_leaves_allowed - leave_allocated
+		if leave_alloc_balance > 0:
+			is_carry_forward = frappe.db.get_value("Leave Type", self.leave_type, "is_carry_forward")
+			print "is_carry_forward: ", is_carry_forward
+			allocation = frappe.new_doc("Leave Allocation")
+			allocation.employee = self.employee
+			allocation.employee_name = self.employee_name
+			allocation.leave_type = self.leave_type
+			allocation.from_date = self.work_from_date
+			allocation.to_date = self.work_end_date
+			allocation.new_leaves_allocated = leave_alloc_balance
+			allocation.total_leaves_allocated = leave_alloc_balance
+			allocation.compensatory_request = self.name
+			allocation.description = self.reason
+			if is_carry_forward == 1:
+				allocation.carry_forward = True
+			allocation.save(ignore_permissions = True)
+			allocation.submit()
+		else:
+			frappe.throw(_("Maximum of leave allocation exceed"))
+
+
+	def get_leave_period(self, company):
+		return frappe.db.sql("""
+			select name, from_date, to_date
+			from `tabLeave Period`
+			where company=%(company)s and is_active=1
+				and (from_date between %(from_date)s and %(to_date)s
+					or to_date between %(from_date)s and %(to_date)s
+					or (from_date < %(from_date)s and to_date > %(to_date)s))
+		""", {
+			"from_date": self.work_from_date,
+			"to_date": self.work_end_date,
+			"company": company
+		}, as_dict=1)
+
+	def validate_dates(self):
+		date_of_joining = frappe.db.get_value("Employee", self.employee, "date_of_joining")
+		if getdate(self.work_from_date) >= getdate(self.work_end_date):
+			frappe.throw(_("Work end date can not be equal or less than from date"))
+		elif date_of_joining and getdate(self.work_from_date) < getdate(date_of_joining):
+			frappe.throw(_("Work from date can not be less than employee's joining date"))
+
+	def validate_overlap(self):
```

Comment:
Seems like overlapping check is a common function in many docs. We can commonify all overlap functions in a single function.